N+1 Problem Resolution and Integration Test Implementation

lily4178993 opened the pull request:
## Description:

In collaboration with my partner @Ayokunnumi1, we have addressed the N+1 problem across various controllers, with a particular focus on resolving the issue in the `posts#index` action. We ensured that all posts and their associated comments are fetched efficiently, eliminating unnecessary database queries.

### N+1 Problem Resolution:

- Ensured efficient loading of all posts and their associated comments for a user.
- Deliberately avoided the use of the Bullet gem for analysis, preferring to inspect the logs for performance improvements.
Below are visual comparisons:
**Before:**
![image](https://github.com/lily4178993/nell-blog/assets/101261047/ec4f6114-270d-4f5d-bc74-52c9a9046dd6)
**After:**
![image](https://github.com/lily4178993/nell-blog/assets/101261047/03ac28a5-60e6-423f-aa6b-0ff6f850c1a5)



### Integration Test Implementation:

We have incorporated Capybara to create comprehensive integration tests for essential views within the project. The tests are designed to validate the functionality and flow of each page, ensuring a seamless user experience.

#### User Index Page:

- Verified that the username of all users is displayed.
- Checked for the presence of profile pictures for each user.
- Confirmed the visibility of the number of posts written by each user.
- Implemented redirection to a user's show page upon clicking on their profile.

#### User Show Page:

- Confirmed the display of the user's profile picture.
- Ensured visibility of the user's username.
- Checked the accuracy of the displayed post count for the user.
- Verified the visibility of the user's bio.
- Confirmed the display of the user's first three posts.
- Implemented a button for viewing all of a user's posts.
- Validated redirection to a post's show page upon clicking a user's post.
- Verified redirection to the user's post index page when viewing all posts.

#### User Post Index Page:

- Confirmed the visibility of the user's profile picture.
- Ensured the display of the user's username.
- Checked the accuracy of the displayed post count for the user.
- Validated the display of a post's title.
- Confirmed visibility of some of the post's body.
- Checked the display of the first comments on a post.
- Verified the accuracy of the displayed comment and like counts.
- Implemented a section for pagination for an optimal viewing experience.
- Validated redirection to a post's show page upon clicking a post.

#### Post Show Page:

- Verified the display of the post's title.
- Ensured visibility of the post's author.
- Checked the accuracy of the displayed comment and like counts.
- Confirmed the display of the post body.
- Checked the visibility of the username for each commenter.
- Validated the display of each comment made on the post.

These changes collectively enhance the project's performance and provide a more robust testing framework for future development.

## Additional Notes:

Please review the changes made and confirm that they align with the intended enhancements and fulfill the specified integration test scenarios. If further adjustments are needed, we are open to feedback for continuous improvement.

lily4178993 committed:
feat: create empty user#index spec file

lily4178993 committed:
feat: Add user post index view spec

lily4178993 committed:
feat: rename posts#index spec file

lily4178993 committed:
test: Add feature spec for Post show page

lily4178993 committed:
feat: Refactor post index view

lily4178993 committed:
test: Refactor post_index_spec.rb to include ActionView::Helpers::TextHelper

lily4178993 committed:
feat: Refactor controller code to include eager loading

Ayokunnumi1 committed:
Remove test for bio in users index spec

Ayokunnumi1 committed:
fix linters error

Ayokunnumi1 committed:
fix linters error

Ayokunnumi1 committed:
Edit user show view file

Ayokunnumi1 committed:
Add test cases for user show spec file

lily4178993 committed:
test: Remove controller spec files

lily4178993 committed:
Update post index query to include author and order by created_at

lily4178993 committed:
feat: Refactor comments retrieval in show action

lily4178993 committed:
feat: Refactor posts controller and user show view

Olamarx commented on spec/views/user_show_spec.rb:
```diff
@@ -0,0 +1,31 @@
+    render
+  end
+
+  it 'displays user details' do
+    expect(rendered).to have_content(user.name)
+    expect(rendered).to have_content(user.bio)
+    expect(rendered).to have_css("img[src*='#{user.photo}']")
+    expect(rendered).to have_content("Number of posts: #{user.posts_counter}")
+  end
+
+  it 'displays user posts' do
+    posts.each do |post|
+      expect(rendered).to have_content(post.title)
+      expect(rendered).to have_content(post.text) # Assuming 'text' is the attribute for post body
+    end
```
- [ ] While you have done a good job so far, your tests for this file are not yet complete, kindly add the pointed ones in the image
![image](https://github.com/lily4178993/nell-blog/assets/86200356/01d6ac74-8cce-48e7-9a4e-e31ac7f7beb4)

Olamarx commented on spec/views/post_index_spec.rb:
```diff
@@ -0,0 +1,44 @@
+
+    expect(page).to have_content("Comments: #{post.comments_counter}")
+    expect(page).to have_content("Likes: #{post.likes_counter}")
+  end
+
+  scenario 'User sees pagination section on index page' do
+    3.times { Post.create(author: user, title: 'Test Post', text: 'This is a test post.') }
+
+    visit user_posts_path(user)
+
+    expect(page).to have_css('.post-container', count: 2)
+  end
+
+  scenario 'User redirects to post show page when clicking on a post' do
```
- [ ] Same for this file, the test is yet to be complete as pointed in the image 
![image](https://github.com/lily4178993/nell-blog/assets/86200356/06ced311-ed0b-498b-8dce-e22270e04a20)
![image](https://github.com/lily4178993/nell-blog/assets/86200356/bbc125ba-3313-4983-bceb-9bf158437b2b)

Olamarx reviewed (changes requested):
Hi @lily4178993, 

Good job so far, indeed you have spent time on this project!
There are some issues that you still need to work on to go to the next project but you are almost there!
<img src="https://media3.giphy.com/media/l2Je9zHYveK012EVi/giphy.gif"/>

## Highlights
- Linters passed ✔️
- Readme file professional ✔️
- N+1 problem fixed ✔️

## Required Changes ♻️
_Check the comments under the review._
- [ ] I noticed some of your tests are failing, kindly fix it up as seen in the image 
![image](https://github.com/lily4178993/nell-blog/assets/86200356/7ad090bd-1153-4aba-aa8f-0efa58780d5a)

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._
Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

lily4178993 committed:
feat: Refactor post_index_spec.rb to include multiple posts and comments

lily4178993 committed:
feat: Refactor user show page spec and add new scenarios

lily4178993 committed:
feat: Update user creation in user_index_spec.rb

lily4178993 commented:
Hi @Olamarx , Thank you for the review and feedbacks.

## Changes made:
- Refactored integration specs files
- Added additonal tests

screenshot of the tests passing:
![image](https://github.com/lily4178993/nell-blog/assets/101261047/227c2119-4503-4481-909c-c29862b85c3b)

omar-muhamad commented on spec/views/post_index_spec.rb:
```diff
@@ -0,0 +1,39 @@
+    expect(page).to have_content(Post.first.text)
+    expect(page).to have_content(Post.first.comments_counter)
+    expect(page).to have_content(Post.first.likes_counter)
+    expect(page).to have_content(Comment.first.text)
+  end
```
- [ ] As the previous reviewer asked, this test should ensure you can see the first comments (more than one comment) for a user post, not only see the first comment as you did. Please fix that.

omar-muhamad commented on spec/views/post_index_spec.rb:
```diff
@@ -0,0 +1,39 @@
+    expect(page).to have_content(Comment.first.text)
+  end
+
+  scenario 'User sees pagination section on index page to see more posts' do
+    expect(page).to have_css('.post-container', count: 2)
+  end
```
- [ ] As the previous reviewer asked, This test should ensure that the user sees the pagination section on the index page to see more posts, not see if the `div` which has a class `post-container` exists. If you want to make it the correct way you should test the existence of an element that has the `pagination-container` class according to your `post_index` view.

![image](https://github.com/lily4178993/nell-blog/assets/68796232/d90778d0-26f4-4ab7-8305-5414f7911ab0)

omar-muhamad reviewed (changes requested):
# Changes Required ♻️
### Hi @lily4178993,

#### Nice job making the required changes. Unfortunately, you missed some of them but you are almost there.

![](https://media.giphy.com/media/Jq824R93JsLwZCaiSL/giphy.gif)

## (Highlights) Good Points: :+1:
- Your work is well documented. ✔
- The correct workflow is used.  ✔
- All linters checks are ok. ✔

## (Changes Required) Aspects to improve: ♻️

- [ ] Please check the inline comments below ⬇️.

## [Optional] suggestions:

- Nothing to mention.

_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

**Cheers and Happy coding!👏👏👏**

Feel free to leave any questions or comments in the PR thread if something is not 100% clear, and please remember to tag me in your question so I can receive the notification**

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

lily4178993 committed:
Refactor post_index_spec.rb to display post comments correctly

thecodechaser reviewed (approved):
## STATUS: APPROVED 🟢

## Hi @lily4178993, and @Ayokunnumi1👋, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## To Highlight 🎉
- Good job fixing n+1 problem. ✔️
- Good job with integration specs. ✔️
- Gitflow is followed. :heavy_check_mark:
- Professional README file. :heavy_check_mark:
- No linter errors. :heavy_check_mark:
- PR has a good title and summary. ✔️

<img src="https://media3.giphy.com/media/g9582DNuQppxC/giphy-downsized-medium.gif"/>

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

lily4178993 commented:
Thank you @Omar-Muhamad, @thecodechaser and @Olamarx for dedicating time to review my pull request. Many thanks for all your valuable feedback.
